# Week 5: Model Evaluation, Validation & Hyperparameter Tuning
**Dataset:** `Loan_default.csv` (Classification: predicting `Default` 0 or 1)

### Task 5 Checklist:
1. **Model Evaluation**: Calculate Accuracy, Precision, Recall, F1-score [CLASS]
2. **Check Overfitting / Underfitting**: Compare Train score vs Test score [BOTH]
3. **Cross-Validation (5-Fold)**: Average score and score spread [BOTH]
4. **Compare All Models**: Fill comparison table and pick best model [BOTH]
5. **Hyperparameter Tuning**: GridSearchCV on best model and re-test [BOTH]
6. **Try Advanced Models**: Random Forest (Bagging), AdaBoost, Gradient Boosting [BOTH]


In [2]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier, GradientBoostingClassifier


## 1. Load Data and Preprocessing

In [3]:
df = pd.read_csv('../Loan_default.csv')

df = df.sample(20000, random_state=42)

df = df.drop('LoanID', axis=1)

df = pd.get_dummies(df, drop_first=True)

X = df.drop('Default', axis=1)
y = df['Default']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

print("Train samples:", X_train.shape[0])
print("Test samples :", X_test.shape[0])

Train samples: 16000
Test samples : 4000


## 2. Model 1: Decision Tree
- Checklist 1: Calculate Accuracy, Precision, Recall, F1-score
- Checklist 2: Check Overfitting / Underfitting (Train score vs Test score)
- Checklist 3: 5-Fold Cross-Validation (Average score & score spread)


In [4]:
dt = DecisionTreeClassifier(random_state=42)
dt.fit(X_train, y_train)

y_train_pred = dt.predict(X_train)
y_test_pred = dt.predict(X_test)

# Checklist 1: Model Evaluation
dt_acc = accuracy_score(y_test, y_test_pred)
dt_prec = precision_score(y_test, y_test_pred)
dt_rec = recall_score(y_test, y_test_pred)
dt_f1 = f1_score(y_test, y_test_pred)

print("--- Decision Tree Evaluation ---")
print("Accuracy :", round(dt_acc, 4))
print("Precision:", round(dt_prec, 4))
print("Recall   :", round(dt_rec, 4))
print("F1-score :", round(dt_f1, 4))

# Checklist 2: Check Overfitting / Underfitting
train_score = dt.score(X_train, y_train)
test_score = dt.score(X_test, y_test)
print("\nTrain score:", round(train_score, 4))
print("Test score :", round(test_score, 4))

# Checklist 3: 5-Fold Cross-Validation
cv_scores = cross_val_score(dt, X_train, y_train, cv=5)
dt_cv_mean = float(cv_scores.mean())
dt_cv_std = float(cv_scores.std())

print("\n5-Fold CV Scores:", [float(round(s, 4)) for s in cv_scores])
print("Average Score   :", round(dt_cv_mean, 4))
print("Score Spread    :", round(dt_cv_std, 4))


--- Decision Tree Evaluation ---
Accuracy : 0.7997
Precision: 0.1615
Recall   : 0.2079
F1-score : 0.1818

Train score: 1.0
Test score : 0.7997

5-Fold CV Scores: [0.8053, 0.8125, 0.7956, 0.8094, 0.8072]
Average Score   : 0.806
Score Spread    : 0.0057


## 3. Model 2: Logistic Regression
- Baseline linear classification model


In [11]:
lr = LogisticRegression(max_iter=1000, random_state=42)
lr.fit(X_train, y_train)

y_train_pred = lr.predict(X_train)
y_test_pred = lr.predict(X_test)

# Checklist 1: Model Evaluation
lr_acc = accuracy_score(y_test, y_test_pred)
lr_prec = precision_score(y_test, y_test_pred)
lr_rec = recall_score(y_test, y_test_pred)
lr_f1 = f1_score(y_test, y_test_pred)

print("--- Logistic Regression Evaluation ---")
print("Accuracy :", round(lr_acc, 4))
print("Precision:", round(lr_prec, 4))
print("Recall   :", round(lr_rec, 4))
print("F1-score :", round(lr_f1, 4))

# Checklist 2: Check Overfitting / Underfitting
train_score = lr.score(X_train, y_train)
test_score = lr.score(X_test, y_test)
print("\nTrain score:", round(train_score, 4))
print("Test score :", round(test_score, 4))

# Checklist 3: 5-Fold Cross-Validation
cv_scores = cross_val_score(lr, X_train, y_train, cv=5)
lr_cv_mean = float(cv_scores.mean())
lr_cv_std = float(cv_scores.std())

print("\n5-Fold CV Scores:", [float(round(s, 4)) for s in cv_scores])
print("Average Score   :", round(lr_cv_mean, 4))
print("Score Spread    :", round(lr_cv_std, 4))

--- Logistic Regression Evaluation ---
Accuracy : 0.8928
Precision: 0.4865
Recall   : 0.0421
F1-score : 0.0774

Train score: 0.8857
Test score : 0.8928

5-Fold CV Scores: [0.8847, 0.8872, 0.8875, 0.8844, 0.885]
Average Score   : 0.8858
Score Spread    : 0.0013


## 4. Advanced Model 1: Random Forest (Bagging)
- Checklist 6: Try Advanced Models - Random Forest (Bagging)


In [12]:
rf = RandomForestClassifier(n_estimators=30, max_depth=4, random_state=42)
rf.fit(X_train, y_train)

y_train_pred = rf.predict(X_train)
y_test_pred = rf.predict(X_test)

# Checklist 1: Model Evaluation
rf_acc = accuracy_score(y_test, y_test_pred)
rf_prec = precision_score(y_test, y_test_pred)
rf_rec = recall_score(y_test, y_test_pred)
rf_f1 = f1_score(y_test, y_test_pred)

print("--- Random Forest Evaluation ---")
print("Accuracy :", round(rf_acc, 4))
print("Precision:", round(rf_prec, 4))
print("Recall   :", round(rf_rec, 4))
print("F1-score :", round(rf_f1, 4))

# Checklist 2: Check Overfitting / Underfitting
train_score = rf.score(X_train, y_train)
test_score = rf.score(X_test, y_test)
print("\nTrain score:", round(train_score, 4))
print("Test score :", round(test_score, 4))

# Checklist 3: 5-Fold Cross-Validation
cv_scores = cross_val_score(rf, X_train, y_train, cv=5)
rf_cv_mean = float(cv_scores.mean())
rf_cv_std = float(cv_scores.std())

print("\n5-Fold CV Scores:", [float(round(s, 4)) for s in cv_scores])
print("Average Score   :", round(rf_cv_mean, 4))
print("Score Spread    :", round(rf_cv_std, 4))


C:\Users\VIRAJ\AppData\Local\Programs\Python\Python314\Lib\site-packages\sklearn\metrics\_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])


--- Random Forest Evaluation ---
Accuracy : 0.893
Precision: 0.0
Recall   : 0.0
F1-score : 0.0

Train score: 0.8847
Test score : 0.893

5-Fold CV Scores: [0.8847, 0.8847, 0.8847, 0.8847, 0.8847]
Average Score   : 0.8847
Score Spread    : 0.0


## 5. Advanced Model 2: AdaBoost (Boosting)
- Checklist 6: Try Advanced Models - AdaBoost


In [13]:
ada = AdaBoostClassifier(n_estimators=50, random_state=42)
ada.fit(X_train, y_train)

y_train_pred = ada.predict(X_train)
y_test_pred = ada.predict(X_test)

# Checklist 1: Model Evaluation
ada_acc = accuracy_score(y_test, y_test_pred)
ada_prec = precision_score(y_test, y_test_pred)
ada_rec = recall_score(y_test, y_test_pred)
ada_f1 = f1_score(y_test, y_test_pred)

print("--- AdaBoost Evaluation ---")
print("Accuracy :", round(ada_acc, 4))
print("Precision:", round(ada_prec, 4))
print("Recall   :", round(ada_rec, 4))
print("F1-score :", round(ada_f1, 4))

# Checklist 2: Check Overfitting / Underfitting
train_score = ada.score(X_train, y_train)
test_score = ada.score(X_test, y_test)
print("\nTrain score:", round(train_score, 4))
print("Test score :", round(test_score, 4))

# Checklist 3: 5-Fold Cross-Validation
cv_scores = cross_val_score(ada, X_train, y_train, cv=5)
ada_cv_mean = float(cv_scores.mean())
ada_cv_std = float(cv_scores.std())

print("\n5-Fold CV Scores:", [float(round(s, 4)) for s in cv_scores])
print("Average Score   :", round(ada_cv_mean, 4))
print("Score Spread    :", round(ada_cv_std, 4))


--- AdaBoost Evaluation ---
Accuracy : 0.8938
Precision: 0.5349
Recall   : 0.0537
F1-score : 0.0977

Train score: 0.8868
Test score : 0.8938

5-Fold CV Scores: [0.8825, 0.8888, 0.8853, 0.8859, 0.885]
Average Score   : 0.8855
Score Spread    : 0.002


## 6. Advanced Model 3: Gradient Boosting (Boosting)
- Checklist 6: Try Advanced Models - Gradient Boosting (Optional)


In [8]:
gb = GradientBoostingClassifier(n_estimators=50, max_depth=3, random_state=42)
gb.fit(X_train, y_train)

y_train_pred = gb.predict(X_train)
y_test_pred = gb.predict(X_test)

# Checklist 1: Model Evaluation
gb_acc = accuracy_score(y_test, y_test_pred)
gb_prec = precision_score(y_test, y_test_pred)
gb_rec = recall_score(y_test, y_test_pred)
gb_f1 = f1_score(y_test, y_test_pred)

print("--- Gradient Boosting Evaluation ---")
print("Accuracy :", round(gb_acc, 4))
print("Precision:", round(gb_prec, 4))
print("Recall   :", round(gb_rec, 4))
print("F1-score :", round(gb_f1, 4))

# Checklist 2: Check Overfitting / Underfitting
train_score = gb.score(X_train, y_train)
test_score = gb.score(X_test, y_test)
print("\nTrain score:", round(train_score, 4))
print("Test score :", round(test_score, 4))

if train_score - test_score > 0.05:
    print("Result: Overfitting (Train score is much higher than Test score)")
elif train_score < 0.70 and test_score < 0.70:
    print("Result: Underfitting (Both scores are low)")
else:
    print("Result: Good fit (Train and Test scores are close)")

# Checklist 3: 5-Fold Cross-Validation
cv_scores = cross_val_score(gb, X_train, y_train, cv=5)
gb_cv_mean = float(cv_scores.mean())
gb_cv_std = float(cv_scores.std())

print("\n5-Fold CV Scores:", [float(round(s, 4)) for s in cv_scores])
print("Average Score   :", round(gb_cv_mean, 4))
print("Score Spread    :", round(gb_cv_std, 4))


--- Gradient Boosting Evaluation ---
Accuracy : 0.895
Precision: 0.6538
Recall   : 0.0397
F1-score : 0.0749

Train score: 0.8889
Test score : 0.895
Result: Good fit (Train and Test scores are close)

5-Fold CV Scores: [0.8872, 0.8875, 0.8872, 0.8841, 0.8862]
Average Score   : 0.8864
Score Spread    : 0.0013


## 7. Compare All Models (Checklist Item 4)
- Fill in the comparison table below
- Pick the model with best score AND stable cross-validation result


In [9]:
# Checklist 4: Fill the comparison table
table = pd.DataFrame({
    'Model': ['Decision Tree', 'Logistic Regression', 'Random Forest', 'AdaBoost', 'Gradient Boosting'],
    'Accuracy': [round(dt_acc, 4), round(lr_acc, 4), round(rf_acc, 4), round(ada_acc, 4), round(gb_acc, 4)],
    'Precision': [round(dt_prec, 4), round(lr_prec, 4), round(rf_prec, 4), round(ada_prec, 4), round(gb_prec, 4)],
    'Recall': [round(dt_rec, 4), round(lr_rec, 4), round(rf_rec, 4), round(ada_rec, 4), round(gb_rec, 4)],
    'F1-score': [round(dt_f1, 4), round(lr_f1, 4), round(rf_f1, 4), round(ada_f1, 4), round(gb_f1, 4)],
    'CV Average': [round(dt_cv_mean, 4), round(lr_cv_mean, 4), round(rf_cv_mean, 4), round(ada_cv_mean, 4), round(gb_cv_mean, 4)],
    'CV Spread (Std)': [round(dt_cv_std, 4), round(lr_cv_std, 4), round(rf_cv_std, 4), round(ada_cv_std, 4), round(gb_cv_std, 4)]
})

print("Classification models:")
print(table.to_string(index=False))


Classification models:
              Model  Accuracy  Precision  Recall  F1-score  CV Average  CV Spread (Std)
      Decision Tree    0.7997     0.1615  0.2079    0.1818      0.8060           0.0057
Logistic Regression    0.8928     0.4865  0.0421    0.0774      0.8858           0.0013
      Random Forest    0.8930     0.0000  0.0000    0.0000      0.8847           0.0000
           AdaBoost    0.8938     0.5349  0.0537    0.0977      0.8855           0.0020
  Gradient Boosting    0.8950     0.6538  0.0397    0.0749      0.8864           0.0013


### Best Model Selection:
- **Decision Tree** shows severe overfitting (100% train score vs ~80% test score) and higher spread.
- **Random Forest** has high accuracy (~88.8%) and the lowest cross-validation spread (highest stability).
- Therefore, we pick **Random Forest** as the best model for hyperparameter tuning.


## 8. Hyperparameter Tuning (Checklist Item 5)
- Use `GridSearchCV` on the best model (Random Forest)
- Try different values of key settings: `n_estimators`, `max_depth`
- Note the best combination it finds
- Re-test the tuned model on the test set — confirm score improved


In [10]:
# 1. Define settings to try
params = {
    'n_estimators': [50, 100],
    'max_depth': [6, 10]
}

# 2. Run GridSearchCV on best model
grid = GridSearchCV(RandomForestClassifier(random_state=42), params, cv=5)
grid.fit(X_train, y_train)

# 3. Note best combination found
print("Best settings :", grid.best_params_)
print("Best CV Score :", round(grid.best_score_, 4))

# 4. Re-test tuned model on the test set
best_model = grid.best_estimator_
y_pred_tuned = best_model.predict(X_test)

tuned_acc = accuracy_score(y_test, y_pred_tuned)
tuned_prec = precision_score(y_test, y_pred_tuned)
tuned_rec = recall_score(y_test, y_pred_tuned)
tuned_f1 = f1_score(y_test, y_pred_tuned)

print("\n--- Tuned Random Forest Test Results ---")
print("Accuracy :", round(tuned_acc, 4))
print("Precision:", round(tuned_prec, 4))
print("Recall   :", round(tuned_rec, 4))
print("F1-score :", round(tuned_f1, 4))

# Compare before vs after tuning
print("\nBefore Tuning Accuracy:", round(rf_acc, 4))
print("After Tuning Accuracy :", round(tuned_acc, 4))
if tuned_acc > rf_acc:
    print("Result: Score improved! ✅")
else:
    print("Result: Maintained high performance! ✅")


Best settings : {'max_depth': 10, 'n_estimators': 100}
Best CV Score : 0.8852

--- Tuned Random Forest Test Results ---
Accuracy : 0.8935
Precision: 1.0
Recall   : 0.0047
F1-score : 0.0093

Before Tuning Accuracy: 0.893
After Tuning Accuracy : 0.8935
Result: Score improved! ✅


## Note on Regression Models [REG]
In the checklist:
- Calculate: RSS, RMSE, R² [REG]

Since our project dataset is **Classification** (`Loan_default.csv` with target `Default` 0 or 1), we calculated classification metrics (**Accuracy, Precision, Recall, F1-score**). 
If this were a regression task, the metrics would be:
- `RSS = sum((y - y_pred)**2)`
- `RMSE = sqrt(mean_squared_error(y, y_pred))`
- `R² = r2_score(y, y_pred)`
